epytope/TCRSpecificityPrediction/External.py

```diff
@@ -709,4 +709,118 @@ def format_results(self, filenames, tcrs, epitopes, pairwise):
+        return df_out
+    
+class DLpTCR(ARepoTCRSpecificityPrediction):
+    """
+    Author: Xu et al.
+    Paper: https://pubmed.ncbi.nlm.nih.gov/34415016/
+    Repo: https://github.com/JiangBioLab/DLpTCR
+    """
+    __name = "DLpTCR"
+    __version = ""
+    __trc_length = (8, 20)
+    __epitope_length = (0, 30) #not sure, 9 for analysis, but no more info
+    __repo = "https://github.com/JiangBioLab/DLpTCR"
+
+    @property
+    def version(self):
+        return self.__version
+
+    @property
+    def name(self):
+        return self.__name
+
+    @property
+    def tcr_length(self):
+        return self.__trc_length
+
+    @property
+    def epitope_length(self):
+        return self.__epitope_length
+
+    @property
+    def repo(self):
+        return self.__repo
+
+    def format_tcr_data(self, tcrs, epitopes, pairwise, **kwargs):
+        rename_columns = {
+            "VJ_cdr3": "TCRA_CDR3",
+            "VDJ_cdr3": "TCRB_CDR3"
+        }
+        required_columns = list(rename_columns.values()) + ["Epitope"]
+        df_tcrs = tcrs.to_pandas(rename_columns=rename_columns)
+        if pairwise:
+            df_tcrs = self.combine_tcrs_epitopes_pairwise(df_tcrs, epitopes)
+        else:
+            df_tcrs = self.combine_tcrs_epitopes_list(df_tcrs, epitopes)
+        model_type = "B" if "model_type" not in kwargs else kwargs["model_type"]
+        if model_type == "B":
+            df_tcrs = self.filter_by_length(df_tcrs, None, "TCRB_CDR3", "Epitope")
+            df_tcrs = df_tcrs[(~df_tcrs["TCRB_CDR3"].isna()) & (df_tcrs["TCRB_CDR3"] != "")]
+            prediction_columns = ["TCRB_CDR3", "Epitope"]
+        elif model_type == "A":
+            df_tcrs = self.filter_by_length(df_tcrs, "TCRA_CDR3", None, "Epitope")
+            df_tcrs = df_tcrs[(~df_tcrs["TCRA_CDR3"].isna()) & (df_tcrs["TCRA_CDR3"] != "")]
+            prediction_columns = ["TCRA_CDR3", "Epitope"]
+        elif model_type == "AB":
+            df_tcrs = self.filter_by_length(df_tcrs, "TCRA_CDR3", "TCRB_CDR3", "Epitope")
+            df_tcrs = df_tcrs[(~df_tcrs["TCRB_CDR3"].isna()) & (df_tcrs["TCRB_CDR3"] != "")]
+            df_tcrs = df_tcrs[(~df_tcrs["TCRA_CDR3"].isna()) & (df_tcrs["TCRA_CDR3"] != "")]
+            prediction_columns = ["TCRB_CDR3", "TCRA_CDR3", "Epitope"]
+        else:
+            raise TypeError(f"Please specify a correct model_type: A, B or AB")
+        df_tcrs = df_tcrs[required_columns]
+        df_tcrs.drop_duplicates(subset=prediction_columns, inplace=True, keep="first")
+        df_tcrs = df_tcrs.reset_index(drop=True)
+        return df_tcrs
+    
+    def save_tmp_files(self, data, **kwargs):
+        tmp_folder = self.get_tmp_folder_path()
+        model_type = "B" if "model_type" not in kwargs else kwargs["model_type"]
+        path_in = os.path.join(tmp_folder.name, f"{self.name}_input.xlsx")
+        path_out = os.path.join(tmp_folder.name, f"TCR{model_type}_pred.csv")
```

Comment:
is is necessary to have the modeltype in the output file? Otherwise, please remove

Reply:
yes, it is necessary